File: client/src/api/admin/site.ts
import { apiGet, apiPatch, apiPostVoid, type ApiRequestOptions } from '../core';
import {
  AdminSiteIo,
  type AdminSite,
  type SendTestEmailRequest,
  type UpdateAdminSiteRequest,
} from './site.io';

/** `GET /admin/site` — full state for the operator editor. */
export async function getAdminSite(
  options: ApiRequestOptions = {},
): Promise<AdminSite> {
  return apiGet('/admin/site', AdminSiteIo, options);
}

/**
 * `PATCH /admin/site` — partial update. Returns the full updated state so the
 * form can refresh its values and the caller can derive the slim public shape
 * to refresh the `useSite()` context.
 *
 * On 422, throws `ValidationError` (from `core`) carrying the per-field
 * messages (`siteName`, `tosMd`, `privacyMd`).
 */
export async function updateAdminSite(
  body: UpdateAdminSiteRequest,
  options: ApiRequestOptions = {},
): Promise<AdminSite> {
  return apiPatch('/admin/site', body, AdminSiteIo, options);
}

/**
 * `POST /admin/site/test-email` — send one message using the settings in the
 * body rather than the stored ones, so a connection can be verified before
 * it's saved. Always 204.
 *
 * On 422, throws `ValidationError` carrying `to`. A missing host or
 * from-address, and any SMTP failure, arrive as `HttpError` with a message
 * worth showing — the point of the button is the reason it failed.
 */
export const sendTestEmail = (
  body: SendTestEmailRequest,
  options: ApiRequestOptions = {},
): Promise<void> => apiPostVoid('/admin/site/test-email', body, options);

File: server/src/routes/admin/site.rs
//! `/admin/site` — operator settings editor. Every endpoint here requires the
//! `MANAGE_SETTINGS` bit. `GET` returns the full state including
//! raw markdown (the form populates its textareas from this);
//! `PATCH` accepts a partial update and returns the updated full
//! state so the FE can refresh its `useSite()` context in-place.

use axum::{
    Json, Router,
    extract::State,
    http::StatusCode,
    routing::{get, post},
};

use crate::{
    AppError, AppState,
    auth::{Identity, require_permission},
    mail::{TestEmailRequest, send_test_email},
    site::{
        AdminSiteDto, UpdateSiteRequest, apply_site_update, fetch_site_admin, validate_site_update,
    },
    user::Permissions,
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/admin/site", get(detail).patch(update))
        .route("/admin/site/test-email", post(test_email))
}

async fn detail(
    State(state): State<AppState>,
    identity: Identity,
) -> Result<Json<AdminSiteDto>, AppError> {
    require_permission(&identity, Permissions::MANAGE_SETTINGS)?;
    fetch_site_admin(state.pool()).await.map(Json)
}

async fn update(
    State(state): State<AppState>,
    identity: Identity,
    Json(req): Json<UpdateSiteRequest>,
) -> Result<Json<AdminSiteDto>, AppError> {
    require_permission(&identity, Permissions::MANAGE_SETTINGS)?;

    let validated = validate_site_update(req).map_err(AppError::Validation)?;
    if validated.is_noop() {
        // Empty body / all-fields-absent: mirror `PATCH /users/me` and
        // reject as a misuse rather than silently re-fetching.
        return Err(AppError::BadRequest(
            "PATCH body must include at least one settable field".into(),
        ));
    }

    apply_site_update(state.pool(), &validated).await?;
    // The name/description ride in every server-rendered <head>; drop the
    // cached copy so the next document request picks the edit up.
    state.clear_site_meta();
    fetch_site_admin(state.pool()).await.map(Json)
}

/// Send a one-off message to prove the relay works. The body carries the
/// editor's unsaved settings, so this can be used before `PATCH` commits them;
/// the stored row contributes only the fallback password and the site name.
async fn test_email(
    State(state): State<AppState>,
    identity: Identity,
    Json(req): Json<TestEmailRequest>,
) -> Result<StatusCode, AppError> {
    require_permission(&identity, Permissions::MANAGE_SETTINGS)?;

    let stored = fetch_site_admin(state.pool()).await?;
    send_test_email(req, &stored).await?;
    Ok(StatusCode::NO_CONTENT)
}
